### ISSUE TITLE
Quadratic slowdown on LSTM models in tflite

### ISSUE THREAD

**Issue by bkirwi** [tensorflow/tensorflow#47804]
**System information**
This reproduces on both colab and my local machine. I assume colab is more standard and thus more interesting, but my local info for completeness:
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, a bit!
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.4.1 (also reproduces on nightly)
- Python version: 3.8.5
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
LSTM models experience quadratic slowdown when converted to tflite.

The attached notebook shows the behaviour clearly: when using a standard keras model for predictions, the prediction latency is linear in the length of the sequence provided, as expected. When the same model is converted to tflite, the performance is initially good but scales up quadratically with the input size, and becomes effectively unusable for large sequences.

**Describe the expected behavior**
I would expect performance to scale linearly with sequence length in tflite, as it does in keras, and as you'd expect from the underlying maths.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1APyFWSL3ln5gZFEotw_5wwUDpaSJrzD-

**Other info / logs**
The notebook shows the context pretty clearly, but I'm happy to provide more info if needed!

**Comment by abattery** [comment]
@renjie-liu could you take a look at this performance experiments?

**Comment by renjie-liu** [comment]
Can you try set fixed input shapes as shown in here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/experimental_new_converter/Keras_LSTM_fusion_Codelab.ipynb

Thanks!

**Comment by bkirwi** [comment]
@renjie-liu - I've updated the notebook, and can confirm that with fixed input shapes it is now linear, and quite a bit faster in absolute terms.

First: thanks very much for the pointer! I believe this unblocks my application.

Is the quadratic behaviour of the unoptimized case expected / is that something that could be fixed?

**Comment by renjie-liu** [comment]
That one seems to be the unfused case.

Haoliang may help take a look? thanks!

**Comment by haozha111** [comment]
The quadratic time was coming from the underlying tensorlist I believe. Currently a TensorListSetItem is written to two concatenation op, it will have O(n) complexity for one single operation (where n is the sequence length). This will be a problem when the sequence length becomes very long in general. And great that fused LSTM is working in this case!

**Comment by bkirwi** [comment]
Would it make sense to keep this open to track the quadratic complexity in the unfused case, or should I let the butler close it?

**Comment by haozha111** [comment]
Hi, we currently don't have plans to optimize the unfused quadratic complexity issue. Are you able to use fused LSTM in tflite instead?

https://groups.google.com/a/tensorflow.org/g/tflite/c/Ub4apUvblN8

**Comment by bkirwi** [comment]
Yeah, I'm personally unblocked for now.

I still think this would be _good_ to fix or document -- the quadratic complexity for this operation is quite surprising, given both the vanilla tf and theoretical behaviour, and it took a while to figure out what was going on -- but I understand that time is limited, and if you don't think you'll get to it it seems fine to close. (And thanks again for the help!)

**Comment by bkirwi** [comment]
Hrm, actually, turns out I am _not_ able to use the fused LSTM... it causes an unacceptable accuracy loss for my problem. (In the example I'm testing out now, the error rate goes from ~0.13 to ~0.30 when I fix the input shape.)

Is that a known issue? If not, happy to open another issue... seems unrelated to the quadratic complexity.

**Comment by abattery** [comment]
You can try flex tensor list op instead with the recent tf nightly version.

```
converter._experimental_lower_tensor_list_ops = False
```

**Comment by renjie-liu** [comment]
TFLite fused LSTM is stateful, so you will need to call `reset_all_variables`: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/python/interpreter.py#L847-L848

between each innovation.

**Comment by bkirwi** [comment]
> TFLite fused LSTM is stateful, so you will need to call reset_all_variables: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/python/interpreter.py#L847-L848
> between each innovation.

Yes, can confirm I'm doing this correctly. (Error rate is closer to 0.7-0.8 without it.)

> You can try flex tensor list op instead with the recent tf nightly version.

I have not tried this! Will report back.

**Comment by bkirwi** [comment]
Hmm, no appreciable difference unfortunately.

```
converter = tf.lite.TFLiteConverter.from_saved_model(tmp)
converter._experimental_lower_tensor_list_ops = True
tflite_model = converter.convert()
```

Is that the right usage?

If it matters for the optimization: my actual model is a stack of multiple bidirectional LSTM layers, followed by a softmax and CTC-decode. Can share code if folks are interested!

**Comment by abattery** [comment]
I would test this with disabling lowering tensor list ops like the following example:

```
converter = tf.lite.TFLiteConverter.from_saved_model(tmp)
converter._experimental_lower_tensor_list_ops = False
tflite_model = converter.convert()
```

**Comment by bkirwi** [comment]
Ah, apologies, I misunderstood... thanks! Performance is improved in that case but still quadratic. I've updated the notebook:
https://colab.research.google.com/drive/1APyFWSL3ln5gZFEotw_5wwUDpaSJrzD-

**Comment by abattery** [comment]
The new performance graph looks like a linear graph.

**Comment by abattery** [comment]
Closing this issue since the performance improvements has been achieved through the Flex tensor list ops.
